cmd/docker/docker_test.go

```diff
@@ -75,3 +86,92 @@ func TestVersion(t *testing.T) {
 	assert.NilError(t, err)
 	assert.Check(t, is.Contains(b.String(), "Docker version"))
 }
+
+func TestFallbackForceExit(t *testing.T) {
+	longRunningCommand := cobra.Command{
+		RunE: func(cmd *cobra.Command, args []string) error {
+			read, _, err := os.Pipe()
+			if err != nil {
+				return err
+			}
+
+			// wait until the parent process sends a signal to exit
+			_, _, err = bufio.NewReader(read).ReadLine()
+			return err
+		},
+	}
+
+	// This is the child process that will run the long running command
+	if os.Getenv("TEST_FALLBACK_FORCE_EXIT") == "1" {
+		fmt.Println("running long command")
+		ctx, cancel := signal.NotifyContext(context.Background(), signals.TerminationSignals...)
+		t.Cleanup(cancel)
+
+		longRunningCommand.SetErr(streams.NewOut(os.Stderr))
+		longRunningCommand.SetOut(streams.NewOut(os.Stdout))
+
+		go forceExitAfter3TerminationSignals(ctx, streams.NewOut(os.Stderr))
+
+		err := longRunningCommand.ExecuteContext(ctx)
+		if err != nil {
+			os.Exit(0)
+		}
+		return
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	t.Cleanup(cancel)
+
+	// spawn the child process
+	cmd := exec.CommandContext(ctx, os.Args[0], "-test.run=TestFallbackForceExit")
```

Comment:
I'm a bit worried about that 🙈 

At this point it should be a `e2e` test as it's no longer a unit test.

I'm fine with leaving that for a follow up though.

Reply:
@thaJeztah  WDYT?

Reply:
i agree that this is more of an e2e test

Reply:
What would be the drawback of this pattern? I understand that this is technically an e2e test, but I think this is a better way than compiling everything into a binary which you should manage, this allows the `go test` command to manage the binary for you.

See the same pattern being used in `os/exec` https://github.com/golang/go/blob/master/src/os/exec/exec_test.go

Reply:
Yeah, also this has A LOT of assumptions on how the tests are actually ran. Depending on the original invocation, this can run more tests than the TestFallbackForceExit (`-test.run` flag is additive).

Reply:
`os/exec` tests are a bit different - the reexec is handled much earlier in the `TestMain`:
https://github.com/golang/go/blob/4f77a83589baa5a3038cc6e35615084dc7e5f0c1/src/os/exec/exec_test.go#L73

Even then, this code is in the Go source and these tests test the `exec` std Go package, so maintainers of that code are probably much more comfortable with having more assumptions about how the tests are executed.

Also, we use `gotestsum` which calls the `go test` for us, and that's also another layer we need to be careful about (not sure what parameters it actually passes to the `go test`).

Reply:
please see https://github.com/docker/cli/pull/5182